torch/csrc/autograd/record_function.cpp

```diff
@@ -137,6 +138,18 @@ bool hasNonSampledCallbacks() {
   return manager().hasNonSampledCallbacks();
 }
 
+void runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
+  TORCH_INTERNAL_ASSERT(
+      rf != nullptr, "Passed in RecordFunction cannot be null.");
+  if (hasCallbacks()) {
```

Comment:
Do we need to worry about thread-safety here? Given the following comment:

> // WARNING: all calls to pushCallback/popCallback are not thread safe and
// must not overlap with other code execution

It seems this could run in multi-thread envs? I saw other `before()` functions didn't consider this either, so maybe it's not relevant. 

Question: When are those callbacks inserted, and usually what are they?

Reply:
An example of when these callbacks are run is when when the autograd profiler is used, `autograd::profiler::enableProfiler()` will push before and end callbacks to be run whenever we record a function. For our case a separate thread will only try to access these callbacks when profiling is enabled and a valid `RecordFunction` exists, so we have the guarantee that the callbacks were already pushed.